src/pool.rs

```diff
@@ -407,83 +419,100 @@ impl ConnectionPool {
 
             // Will return error if timestamp is greater than current system time, which it should never be set to
-            let require_healthcheck = server.last_activity().elapsed().unwrap().as_millis()
-                > self.settings.healthcheck_delay as u128;
+            let require_healthcheck = force_healthcheck
+                || server.last_activity().elapsed().unwrap().as_millis()
+                    > self.settings.healthcheck_delay as u128;
 
             // Do not issue a health check unless it's been a little while
             // since we last checked the server is ok.
             // Health checks are pretty expensive.
             if !require_healthcheck {
+                self.stats.checkout_time(
+                    now.elapsed().as_micros(),
+                    client_process_id,
+                    server.server_id(),
+                );
                 self.stats
-                    .checkout_time(now.elapsed().as_micros(), process_id, server.server_id());
-                self.stats.server_active(process_id, server.server_id());
+                    .server_active(client_process_id, server.server_id());
                 return Ok((conn, address.clone()));
             }
 
-            debug!("Running health check on server {:?}", address);
-
-            self.stats.server_tested(server.server_id());
-
-            match tokio::time::timeout(
-                tokio::time::Duration::from_millis(self.settings.healthcheck_timeout),
-                server.query(";"), // Cheap query as it skips the query planner
-            )
-            .await
+            if self
+                .run_health_check(address, server, now, client_process_id)
+                .await
             {
-                // Check if health check succeeded.
-                Ok(res) => match res {
-                    Ok(_) => {
-                        self.stats.checkout_time(
-                            now.elapsed().as_micros(),
-                            process_id,
-                            conn.server_id(),
-                        );
-                        self.stats.server_active(process_id, conn.server_id());
-                        return Ok((conn, address.clone()));
-                    }
-
-                    // Health check failed.
-                    Err(err) => {
-                        error!(
-                            "Banning instance {:?} because of failed health check, {:?}",
-                            address, err
-                        );
+                return Ok((conn, address.clone()));
+            } else {
+                continue;
+            }
+        }
 
-                        // Don't leave a bad connection in the pool.
-                        server.mark_bad();
+        Err(Error::AllServersDown)
+    }
 
-                        self.ban(address, process_id);
-                        continue;
-                    }
-                },
+    async fn run_health_check(
```

Comment:
Why did you make this into a function if it's only used once?

Reply:
I believe we'll likely want to call this function in other situations in the future, no strong reason though just trying to break the larger function into components